blockchain/chain_test.go

```diff
@@ -1311,3 +1313,381 @@ func TestIsAncestor(t *testing.T) {
 			branch2Nodes[0].hash.String())
 	}
 }
+
+func TestInvalidateBlock(t *testing.T) {
+	tests := []struct {
+		name     string
+		chainGen func() (*BlockChain, []*chainhash.Hash, func())
+	}{
+		{
+			name: "One branch, invalidate once",
+			chainGen: func() (*BlockChain, []*chainhash.Hash, func()) {
+				source := rand.NewSource(time.Now().UnixNano())
+				rand := rand.New(source)
+
+				chain, params, tearDown := utxoCacheTestChain("TestInvalidateBlock-one-branch-invalidate-once")
+				tip := btcutil.NewBlock(params.GenesisBlock)
+
+				// Create block at height 1.
+				var emptySpendableOuts []*testhelper.SpendableOut
+				b1, spendableOuts1, err := addBlock(chain, tip, emptySpendableOuts)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				var allSpends []*testhelper.SpendableOut
+				nextBlock := b1
+				nextSpends := spendableOuts1
+
+				var invalidateHash *chainhash.Hash
+
+				// Create a chain with 11 blocks.
+				for b := 0; b < 10; b++ {
+					newBlock, newSpendableOuts, err := addBlock(chain, nextBlock, nextSpends)
+					if err != nil {
+						t.Fatal(err)
+					}
+
+					nextBlock = newBlock
+
+					if newBlock.Height() == 5 {
+						invalidateHash = newBlock.Hash()
+					}
+
+					allSpends = append(allSpends, newSpendableOuts...)
+
+					var nextSpendsTmp []*testhelper.SpendableOut
+					for i := 0; i < len(allSpends); i++ {
+						randIdx := rand.Intn(len(allSpends))
+
+						spend := allSpends[randIdx]                                       // get
+						allSpends = append(allSpends[:randIdx], allSpends[randIdx+1:]...) // delete
+						nextSpendsTmp = append(nextSpendsTmp, spend)
+					}
+					nextSpends = nextSpendsTmp
+
+					if b%10 == 0 {
+						// Commit the two base blocks to DB
+						if err := chain.FlushUtxoCache(FlushRequired); err != nil {
+							t.Fatalf("unexpected error while flushing cache: %v", err)
+						}
+					}
+				}
+
+				return chain, []*chainhash.Hash{invalidateHash}, tearDown
+			},
+		},
+		{
+			name: "invalidate twice",
+			chainGen: func() (*BlockChain, []*chainhash.Hash, func()) {
+				source := rand.NewSource(0)
+				rand := rand.New(source)
+
+				chain, params, tearDown := utxoCacheTestChain("TestInvalidateBlock-invalidate-twice")
+				tip := btcutil.NewBlock(params.GenesisBlock)
+
+				// Create block at height 1.
+				var emptySpendableOuts []*testhelper.SpendableOut
+				b1, spendableOuts1, err := addBlock(chain, tip, emptySpendableOuts)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				var allSpends []*testhelper.SpendableOut
+				nextBlock := b1
+				nextSpends := spendableOuts1
+
+				var invalidateHash *chainhash.Hash
+
+				// Create a chain with 11 blocks.
+				for b := 0; b < 10; b++ {
```

Comment:
Can consolidate into a helper func based on the other instance above where we make the same 11 block chain.

Reply:
Will finish this promptly

Reply:
Addressed in the latest push